KAFKA-18932: Removed usage of partition max bytes from share fetch requests

adixitconfluent opened the pull request:
### What
This PR aims to remove the usage of partition max bytes from share fetch requests. Partition Max Bytes is being defined by `PartitionMaxBytesStrategy` which was added to the broker as part of PR https://github.com/apache/kafka/pull/17870

### Testing
The code has been tested with the help of already present unit and integration tests.

adixitconfluent committed:
Removed usage of partition max bytes from share fetch requests on the broker

adixitconfluent committed:
Removed spotless checkstyle failure

adixitconfluent committed:
Minor refactor

apoorvmittal10 commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -3139,15 +3135,15 @@ class KafkaApis(val requestChannel: RequestChannel,
     erroneousAndValidPartitionData.erroneous.forEach { (topicIdPartition, partitionData) => erroneous.put(topicIdPartition, partitionData) }
 
-    val interestedWithMaxBytes = new util.LinkedHashMap[TopicIdPartition, Integer]
+    val interestedTopicPartitions = new util.LinkedHashSet[TopicIdPartition]
```
Do we also perform search actively on TopicIdPartition or just iterate over them? If it's generally iteration then wouldn't it be better to align all code on List rather than on Set?

adixitconfluent replied:
At some places, I did find usage for `.contains`, hence I've used `Set`

apoorvmittal10 replied:
Hmmm, can you point me where all we do that, so I can look properly. In initial search on the PR, I can't find.

adixitconfluent replied:
my bad, I saw this [line](https://github.com/apache/kafka/pull/19148/files#diff-cc056b4960ededba37a438b1454f0f3c5ff5e8ad5e6d2ec9a08e813ca056ffebR3000) and a couple of lines in tests because of which I though `Set` is required here, should we use `List` then with an internal implementation of `ArrayList` then?

apoorvmittal10 replied:
I think so that should be good enough. I was just thinking about how easy it would be to rotate ArrayList vs LinkedList as that's the first thing we do.

apoorvmittal10 replied:
Also just check ImpliciLinkedHashCollection in case you find the implementation of same better. Then this can be searchable as well.

adixitconfluent replied:
We'll have to use `TopicIdPartition` as the element type of `ImplicitLinkedHashCollection`, and to do that we'll have to make `TopicIdPartition` implement `ImplicitLinkedHashCollection.Element` which doesn't look ideal. Between ArrayList and LinkedList, I am thinking to use ArrayList because it is better in storing the data and accessing it. It takes lesser memory. Although its insertion and deletion operations are not that efficient, we do not use delete operation anywhere in the code as such and insertion is much rare compared to accessing the list. Please let me know if we can finalize ArrayList.

apoorvmittal10 reviewed (commented):
Thanks for the PR. One high level query before I proceed for review.

adixitconfluent committed:
Addressed Apoorv'r round 1 review comment

apoorvmittal10 commented on clients/src/main/java/org/apache/kafka/common/requests/ShareFetchRequest.java:
```diff
@@ -151,7 +149,7 @@ public String toString() {
 
     private final ShareFetchRequestData data;
-    private volatile LinkedHashMap<TopicIdPartition, ShareFetchRequest.SharePartitionData> shareFetchData = null;
+    private volatile ArrayList<TopicIdPartition> shareFetchData = null;
```
```suggestion
    private volatile List<TopicIdPartition> shareFetchData = null;
```

apoorvmittal10 replied:
Is there a reason to use ArrayList specifically and not declaring the `List` interface.

apoorvmittal10 commented on clients/src/main/java/org/apache/kafka/common/requests/ShareFetchRequest.java:
```diff
@@ -226,23 +189,18 @@ public int maxWait() {
                     // Assigning the lazy-initialized `shareFetchData` in the last step
                     // to avoid other threads accessing a half-initialized object.
-                    final LinkedHashMap<TopicIdPartition, ShareFetchRequest.SharePartitionData> shareFetchDataTmp = new LinkedHashMap<>();
+                    final ArrayList<TopicIdPartition> shareFetchDataTmp = new ArrayList<>();
```
```suggestion
                    final List<TopicIdPartition> shareFetchDataTmp = new ArrayList<>();
```

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartitionManager.java:
```diff
@@ -258,14 +257,14 @@ public CompletableFuture<Map<TopicIdPartition, PartitionData>> fetchMessages(
         int sessionEpoch,
         int batchSize,
-        LinkedHashMap<TopicIdPartition, Integer> partitionMaxBytes
+        ArrayList<TopicIdPartition> topicPartitions
```
```suggestion
        List<TopicIdPartition> topicIdPartitions
```

apoorvmittal10 replied:
Is it done, I don't see the changes?

adixitconfluent replied:
forgot to change the variable name. I only changed the data type. Just pushed the variable name change as well.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartitionManager.java:
```diff
@@ -427,29 +426,20 @@ private CompletableFuture<Map<TopicIdPartition, ShareAcknowledgeResponseData.Par
         if (reqMetadata.isFull()) {
             ShareSessionKey key = shareSessionKey(groupId, reqMetadata.memberId());
             if (reqMetadata.epoch() == ShareRequestMetadata.FINAL_EPOCH) {
-                // If the epoch is FINAL_EPOCH, don't try to create a new session.
-                if (!shareFetchDataWithMaxBytes.isEmpty()) {
-                    throw Errors.INVALID_REQUEST.exception();
-                }
```
Shouldn't we have a similar check i.e. if topic partitions are included in fetch request with final epoch?

adixitconfluent replied:
when we get a `ShareFetch` requests with a `FINAL_EPOCH`, it **can** ideally have some topic partitions only for acknowledgement purpose. 
In the current code, the `FINAL_EPOCH` request is always a `ShareAcknowledge` request. However, I had a discussion with @AndrewJSchofield, he mentioned that for the future, if any dev decides to have a `ShareFetch` request with a `FINAL_EPOCH`, we should be able to support that. In the current code, if we receive a `ShareFetch` request with some topic partitions, we will be only doing acknowledgements for those topics partitions and we won't be doing any fetch for those topic partitions. I have verified this behaviour.

apoorvmittal10 replied:
So with this change a final_epoch can also fetch data?

adixitconfluent replied:
no we cannot fetch data as of now, but `ShareFetch` requests can acknowledge data correctly if sent with `FINAL_EPOCH`

apoorvmittal10 replied:
Sounds good.

apoorvmittal10 commented on core/src/test/java/kafka/server/share/DelayedShareFetchTest.java:
```diff
@@ -134,7 +132,7 @@ public void testDelayedShareFetchTryCompleteReturnsFalseDueToNonAcquirablePartit
 
         ShareFetch shareFetch = new ShareFetch(FETCH_PARAMS, groupId, Uuid.randomUuid().toString(),
-            new CompletableFuture<>(), partitionMaxBytes, BATCH_SIZE, MAX_FETCH_RECORDS,
+            new CompletableFuture<>(), arrayList(tp0, tp1), BATCH_SIZE, MAX_FETCH_RECORDS,
```
```suggestion
            new CompletableFuture<>(), List.of(tp0, tp1), BATCH_SIZE, MAX_FETCH_RECORDS,
```

apoorvmittal10 commented on core/src/test/java/kafka/server/share/DelayedShareFetchTest.java:
```diff
@@ -526,7 +518,7 @@ public void testForceCompleteTriggersDelayedActionsQueue() {
         TopicIdPartition tp1 = new TopicIdPartition(topicId, new TopicPartition("foo", 1));
         TopicIdPartition tp2 = new TopicIdPartition(topicId, new TopicPartition("foo", 2));
-        LinkedHashMap<TopicIdPartition, Integer> partitionMaxBytes1 = orderedMap(PARTITION_MAX_BYTES, tp0, tp1);
+        ArrayList<TopicIdPartition> topicIdPartitions1 = arrayList(tp0, tp1);
```
```suggestion
        List<TopicIdPartition> topicIdPartitions1 = List.of(tp0, tp1);
```

apoorvmittal10 commented on server/src/main/java/org/apache/kafka/server/share/fetch/PartitionRotateStrategy.java:
```diff
@@ -64,8 +63,8 @@ static PartitionRotateStrategy type(StrategyType type) {
      */
-    static LinkedHashMap<TopicIdPartition, Integer> rotateRoundRobin(
-        LinkedHashMap<TopicIdPartition, Integer> topicIdPartitions,
+    static ArrayList<TopicIdPartition> rotateRoundRobin(
```
```suggestion
    static List<TopicIdPartition> rotateRoundRobin(
```

apoorvmittal10 commented on server/src/main/java/org/apache/kafka/server/share/fetch/PartitionRotateStrategy.java:
```diff
@@ -48,7 +47,7 @@ public String toString() {
      * @return the rotated topicIdPartitions
      */
-    LinkedHashMap<TopicIdPartition, Integer> rotate(LinkedHashMap<TopicIdPartition, Integer> topicIdPartitions, PartitionRotateMetadata metadata);
+    ArrayList<TopicIdPartition> rotate(ArrayList<TopicIdPartition> topicIdPartitions, PartitionRotateMetadata metadata);
```
```suggestion
    List<TopicIdPartition> rotate(List<TopicIdPartition> topicIdPartitions, PartitionRotateMetadata metadata);
```

apoorvmittal10 commented on server/src/main/java/org/apache/kafka/server/share/fetch/PartitionRotateStrategy.java:
```diff
@@ -80,20 +79,18 @@ static LinkedHashMap<TopicIdPartition, Integer> rotateRoundRobin(
+        for (TopicIdPartition topicIdPartition : topicIdPartitions) {
             if (i < rotateAt) {
-                suffixPartitions.put(entry.getKey(), entry.getValue());
+                suffixPartitions.add(topicIdPartition);
             } else {
-                rotatedPartitions.put(entry.getKey(), entry.getValue());
+                rotatedPartitions.add(topicIdPartition);
             }
             i++;
         }
-        rotatedPartitions.putAll(suffixPartitions);
+        rotatedPartitions.addAll(suffixPartitions);
```
Should we use `Collections.rotate` instead?

apoorvmittal10 commented on server/src/main/java/org/apache/kafka/server/share/session/ShareSession.java:
```diff
@@ -110,25 +109,20 @@ public synchronized LastUsedKey lastUsedKey() {
         return new LastUsedKey(key, lastUsedMs);
     }
 
-    // Visible for testing
-    public synchronized long creationMs() {
-        return creationMs;
-    }
```
Why we have removed the creationMs, is it not being used?

adixitconfluent replied:
yes

apoorvmittal10 commented on server/src/main/java/org/apache/kafka/server/share/session/ShareSession.java:
```diff
@@ -110,25 +109,20 @@ public synchronized LastUsedKey lastUsedKey() {
-            ShareFetchRequest.SharePartitionData> shareFetchData, List<TopicIdPartition> toForget) {
+    public synchronized Map<ModifiedTopicIdPartitionType, List<TopicIdPartition>> update(
+        List<TopicIdPartition> shareFetchData,
+        List<TopicIdPartition> toForget) {
```
```suggestion
        List<TopicIdPartition> toForget
  ) {
```

apoorvmittal10 commented on server/src/test/java/org/apache/kafka/server/share/fetch/ShareFetchTestUtils.java:
```diff
@@ -30,34 +30,30 @@
-            map.put(tp, partitionMaxBytes);
-        }
-        return map;
+    public static ArrayList<TopicIdPartition> arrayList(TopicIdPartition... topicIdPartitions) {
+        ArrayList<TopicIdPartition> list = new ArrayList<>();
+        Collections.addAll(list, topicIdPartitions);
+        return list;
     }
```
Why not to use List.of()? Do we need mutable list?

apoorvmittal10 reviewed (commented):
Thanks for the PR. Some comments.

adixitconfluent committed:
Addressed Apoorv's round 1 review comments

adixitconfluent committed:
Addressed Apoorv's round 1 review comments

adixitconfluent committed:
Minor refactor

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/common/requests/ShareFetchRequest.java:
```diff
@@ -226,23 +189,18 @@ public int maxWait() {
-                                            shareFetchPartition.partitionMaxBytes()
-                                    )
-                            );
+                            shareFetchDataTmp.add(new TopicIdPartition(shareFetchTopic.topicId(), new TopicPartition(name, shareFetchPartition.partitionIndex())));
```
nit: This could be `new TopicIdPartition(shareFetchTopic.topicId(), shareFetchPartition.partitionIndex(), name)`.

AndrewJSchofield commented on core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:
```diff
@@ -3918,10 +3918,8 @@ class KafkaApisTest extends Logging {
-          new ShareFetchRequest.SharePartitionData(topicId, partitionMaxBytes)
-      ).asJava)
+      new ShareSessionContext(new ShareRequestMetadata(memberId, shareSessionEpoch), List(
+        new TopicIdPartition(topicId, new TopicPartition(topicName, partitionIndex))).asJava)
```
nit: You could use the constructor `TopicIdPartition(Uuid, int, String)` which would tidy the code up a bit.

adixitconfluent replied:
done at all the places wherever the changes were made as part of the PR.

AndrewJSchofield commented on server/src/test/java/org/apache/kafka/server/share/fetch/ShareFetchTestUtils.java:
```diff
@@ -68,27 +50,22 @@ public static LinkedHashMap<TopicIdPartition, Integer> orderedMap(int partitionM
      * @param rotationAt The position to rotate the keys at.
      */
     public static void validateRotatedMapEquals(
```
nit: You're rotating a list not a map now, so the method name is inaccurate.

AndrewJSchofield reviewed (changes requested):
Thanks for the PR. It's good to see some of the cruft being removed.

AndrewJSchofield commented:
Please update `clients/src/main/resources/common/message/ShareFetchRequest.json` too to remove `PartitionMaxBytes`.

adixitconfluent committed:
Addresses Andrew's comments

adixitconfluent commented:
> Please update `clients/src/main/resources/common/message/ShareFetchRequest.json` too to remove `PartitionMaxBytes`.

done this in my latest commit, it has triggered a whole bunch of changes :)

apoorvmittal10 commented on core/src/test/java/kafka/server/share/DelayedShareFetchTest.java:
```diff
@@ -1163,7 +1144,7 @@ public void testPartitionMaxBytesFromUniformStrategyInCombineLogReadResponse() {
         ShareFetch shareFetch = new ShareFetch(
             FETCH_PARAMS, "grp", Uuid.randomUuid().toString(),
-            new CompletableFuture<>(), new LinkedHashMap<>(), BATCH_SIZE, MAX_FETCH_RECORDS,
+            new CompletableFuture<>(), new ArrayList<>(), BATCH_SIZE, MAX_FETCH_RECORDS,
```
```suggestion
            new CompletableFuture<>(), List.of(), BATCH_SIZE, MAX_FETCH_RECORDS,
```

apoorvmittal10 commented on core/src/test/java/kafka/server/share/SharePartitionManagerTest.java:
```diff
@@ -201,17 +198,15 @@ public void testNewContextReturnsFinalContextWithoutRequestData() {
 
         ShareRequestMetadata reqMetadata2 = new ShareRequestMetadata(memberId, ShareRequestMetadata.FINAL_EPOCH);
-        ShareFetchContext context2 = sharePartitionManager.newContext(groupId, Collections.emptyMap(), Collections.emptyList(), reqMetadata2, true);
+        ShareFetchContext context2 = sharePartitionManager.newContext(groupId, Collections.emptyList(), Collections.emptyList(), reqMetadata2, true);
```
```suggestion
        ShareFetchContext context2 = sharePartitionManager.newContext(groupId, List.of(), List.of(), reqMetadata2, true);
```

apoorvmittal10 commented on core/src/test/java/kafka/server/share/SharePartitionManagerTest.java:
```diff
@@ -365,14 +345,14 @@ public void testNewContext() {
 
         // Test generating a throttled response for a subsequent share session
-        ShareFetchContext context7 = sharePartitionManager.newContext(groupId, Collections.emptyMap(), EMPTY_PART_LIST,
+        ShareFetchContext context7 = sharePartitionManager.newContext(groupId, Collections.emptyList(), EMPTY_PART_LIST,
```
```suggestion
        ShareFetchContext context7 = sharePartitionManager.newContext(groupId, List.of(), EMPTY_PART_LIST,
```

apoorvmittal10 commented on core/src/test/java/kafka/server/share/SharePartitionManagerTest.java:
```diff
@@ -365,14 +345,14 @@ public void testNewContext() {
 
         // Get the final share session.
-        ShareFetchContext context8 = sharePartitionManager.newContext(groupId, Collections.emptyMap(), EMPTY_PART_LIST,
+        ShareFetchContext context8 = sharePartitionManager.newContext(groupId, Collections.emptyList(), EMPTY_PART_LIST,
```
```suggestion
        ShareFetchContext context8 = sharePartitionManager.newContext(groupId, List.of(), EMPTY_PART_LIST,
```

apoorvmittal10 commented on core/src/test/java/kafka/server/share/SharePartitionManagerTest.java:
```diff
@@ -458,7 +434,7 @@ public void testShareSessionExpiration() {
 
         // Create a subsequent share fetch context for session 1
-        ShareFetchContext session1context2 = sharePartitionManager.newContext(groupId, Collections.emptyMap(), EMPTY_PART_LIST,
+        ShareFetchContext session1context2 = sharePartitionManager.newContext(groupId, Collections.emptyList(), EMPTY_PART_LIST,
```
```suggestion
        ShareFetchContext session1context2 = sharePartitionManager.newContext(groupId, List.of(), EMPTY_PART_LIST,
```

apoorvmittal10 replied:
And at other places changed in the PR.

apoorvmittal10 commented on core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:
```diff
@@ -4044,7 +4039,6 @@ class KafkaApisTest extends Logging {
             .setPartitionIndex(0)
-            .setPartitionMaxBytes(partitionMaxBytes)
         ).asJava)
       ).asJava)
```
```
shareFetchRequestData = new ShareFetchRequestData().
      setGroupId(groupId).
      setMemberId(memberId.toString).
      setShareSessionEpoch(-1).
      setTopics(util.List.of(new ShareFetchRequestData.FetchTopic().
        setTopicId(topicId).
        setPartitions(util.List.of(
          new ShareFetchRequestData.FetchPartition()
            .setPartitionIndex(0)
            .setAcknowledgementBatches(util.List.of(
              new AcknowledgementBatch()
                .setFirstOffset(0)
                .setLastOffset(9)
                .setAcknowledgeTypes(Collections.singletonList(1.toByte))
            ))
        ))
      ))

adixitconfluent replied:
made the change from `List` to `util.list.of`

apoorvmittal10 commented on core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:
```diff
@@ -4095,9 +4089,8 @@ class KafkaApisTest extends Logging {
-          new ShareFetchRequest.SharePartitionData(topicId, partitionMaxBytes)
+      new ShareSessionContext(new ShareRequestMetadata(memberId, 0), List(
+        new TopicIdPartition(topicId, partitionIndex, topicName)
       ).asJava)
```
Use util.List.of

apoorvmittal10 replied:
And at other places below.

apoorvmittal10 commented on core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:
```diff
@@ -6083,8 +6083,8 @@ class ReplicaManagerTest {
-      val partitionMaxBytes = new util.LinkedHashMap[TopicIdPartition, Integer]
-      partitionMaxBytes.put(tp1, 1000)
+      val topicPartitions = new util.ArrayList[TopicIdPartition]
+      topicPartitions.add(tp1)
```
Do we need mutable list? Else we can use `util.List.of`.

apoorvmittal10 commented on core/src/test/scala/unit/kafka/server/ShareFetchAcknowledgeRequestTest.scala:
```diff
@@ -59,7 +59,7 @@ class ShareFetchAcknowledgeRequestTest(cluster: ClusterInstance) extends GroupCo
     )
 
-    val shareFetchRequest = createShareFetchRequest(groupId, metadata, MAX_PARTITION_BYTES, send, Seq.empty, Map.empty)
+    val shareFetchRequest = createShareFetchRequest(groupId, metadata, send, Seq.empty, Map.empty)
```
Remove `MAX_PARTITION_BYTES` constant in the file.

apoorvmittal10 commented on server/src/main/java/org/apache/kafka/server/share/ErroneousAndValidPartitionData.java:
```diff
@@ -20,47 +20,48 @@
-        shareFetchData.forEach((topicIdPartition, sharePartitionData) -> {
+        validTopicIdPartitions = new ArrayList<>();
+        shareFetchData.forEach(topicIdPartition -> {
             if (topicIdPartition.topic() == null) {
                 erroneous.put(topicIdPartition, ShareFetchResponse.partitionResponse(topicIdPartition, Errors.UNKNOWN_TOPIC_ID));
```
Question, we are checking topic name here and filling error as `UNKNOWN_TOPIC_ID`, is it correct?

adixitconfluent replied:
yes

apoorvmittal10 replied:
Why? The problem is that topic name doesn't exist but we reply with that topic id is unknown, isn't it incorrect?

adixitconfluent replied:
If we are unable to resolve a topic name using its topic id, we give out this error. IIRC, there is a `MetadataCache` object through which  we get `topicIdsToNames` map. We check it from there if the topic id exists.
You can find similar code in the handling of regular fetch request [here](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/KafkaApis.scala#L557) and [here](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/KafkaApis.scala#L574).

apoorvmittal10 commented on server/src/main/java/org/apache/kafka/server/share/ErroneousAndValidPartitionData.java:
```diff
@@ -20,47 +20,48 @@
 
     public ErroneousAndValidPartitionData() {
         this.erroneous = new HashMap<>();
-        this.validTopicIdPartitions = new HashMap<>();
+        this.validTopicIdPartitions = new ArrayList<>();
```
Do we modify `erroneous` or `validTopicIdPartitions` after initializing using default constructor. I am thinking is it correct to have mutable variables?

adixitconfluent replied:
you're right, the default constructor `ErroneousAndValidPartitionData()` is only utilized in `FinalContext` and we don't modify the `erroneous` or `validTopicIdPartitions` post initialization. We only iterate over it. Hence, changed the code to -
```
public ErroneousAndValidPartitionData() {
        this.erroneous = Map.of();
        this.validTopicIdPartitions = List.of();
    }
```

apoorvmittal10 commented on server/src/main/java/org/apache/kafka/server/share/fetch/PartitionRotateStrategy.java:
```diff
@@ -80,20 +81,8 @@ static LinkedHashMap<TopicIdPartition, Integer> rotateRoundRobin(
-        }
-        rotatedPartitions.putAll(suffixPartitions);
+        List<TopicIdPartition> rotatedPartitions = new ArrayList<>(topicIdPartitions);
+        Collections.rotate(rotatedPartitions, -1 * rotateAt);
```
I think for elements [1,2,3] and rotate at 1 we want the output as [2,3,1] instead of [3,1,2] hence you negated. Can you please write comment regarding why `-1`?

adixitconfluent replied:
I have added the comment `We want the elements from the end of the list to move left by the distance provided i.e. if the original list is [1,2,3], and we want to rotate it by 1, we want the output as [2,3,1] and not [3,1,2]. Hence, we need negation of distance here` in the code. Regarding why I multiplied -1 to `rotateAt` instead of `-rotateAt`, its because `-1 * rotateAt` looks cleaner code to me.

apoorvmittal10 commented on server/src/main/java/org/apache/kafka/server/share/fetch/PartitionRotateStrategy.java:
```diff
@@ -80,20 +81,8 @@ static LinkedHashMap<TopicIdPartition, Integer> rotateRoundRobin(
-            i++;
-        }
-        rotatedPartitions.putAll(suffixPartitions);
+        List<TopicIdPartition> rotatedPartitions = new ArrayList<>(topicIdPartitions);
```
Also please write comment that we don't want to modify the input list hence created a copy.

apoorvmittal10 commented on server/src/test/java/org/apache/kafka/server/share/fetch/PartitionRotateStrategyTest.java:
```diff
@@ -35,64 +36,64 @@ public class PartitionRotateStrategyTest {
     public void testRoundRobinStrategy() {
         PartitionRotateStrategy strategy = PartitionRotateStrategy.type(StrategyType.ROUND_ROBIN);
-        LinkedHashMap<TopicIdPartition, Integer> partitions = createPartitions(3);
+        ArrayList<TopicIdPartition> partitions = createPartitions(3);
```
```suggestion
        List<TopicIdPartition> partitions = createPartitions(3);
```

apoorvmittal10 commented on server/src/test/java/org/apache/kafka/server/share/fetch/PartitionRotateStrategyTest.java:
```diff
@@ -35,64 +36,64 @@ public class PartitionRotateStrategyTest {
      */
-    private LinkedHashMap<TopicIdPartition, Integer> createPartitions(int size) {
-        LinkedHashMap<TopicIdPartition, Integer> partitions = new LinkedHashMap<>();
+    private ArrayList<TopicIdPartition> createPartitions(int size) {
```
```suggestion
    private List<TopicIdPartition> createPartitions(int size) {
```

apoorvmittal10 commented on server/src/test/java/org/apache/kafka/server/share/fetch/PartitionRotateStrategyTest.java:
```diff
@@ -35,64 +36,64 @@ public class PartitionRotateStrategyTest {
-    private LinkedHashMap<TopicIdPartition, Integer> createPartitions(int size) {
-        LinkedHashMap<TopicIdPartition, Integer> partitions = new LinkedHashMap<>();
+    private ArrayList<TopicIdPartition> createPartitions(int size) {
+        ArrayList<TopicIdPartition> partitions = new ArrayList<>();
```
```suggestion
        List<TopicIdPartition> partitions = new ArrayList<>();
```

apoorvmittal10 reviewed (commented):
LGTM, left some comments.

adixitconfluent committed:
Addressed Apoorv's review comments round 2 - part 1

adixitconfluent committed:
Addressed Apoorv's review comments round 2 - part 2

adixitconfluent commented:
@apoorvmittal10 , I have addressed your comments. Please review when you can. Also, I had to do a force push to fix the branch history because it got corrupted by a merge commit performed by IntelliJ. That's why you see so many labels in the PR. If possible, can you remove the extra labels `streams`, `producer`, `tools`, `connect`, `kraft`, `mirror-maker-2`, `storage`, `tiered-storage`, `build` and `generator`? 
Thanks!

apoorvmittal10 commented on server/src/main/java/org/apache/kafka/server/share/fetch/PartitionRotateStrategy.java:
```diff
@@ -80,20 +81,11 @@ static LinkedHashMap<TopicIdPartition, Integer> rotateRoundRobin(
-            i++;
-        }
-        rotatedPartitions.putAll(suffixPartitions);
+        // We don't want to modify the original list, hence created a copy.
```
```suggestion
        // Avoid modifying the original list, create copy.
```

apoorvmittal10 commented on server/src/main/java/org/apache/kafka/server/share/fetch/PartitionRotateStrategy.java:
```diff
@@ -80,20 +81,11 @@ static LinkedHashMap<TopicIdPartition, Integer> rotateRoundRobin(
-        rotatedPartitions.putAll(suffixPartitions);
+        // We don't want to modify the original list, hence created a copy.
+        List<TopicIdPartition> rotatedPartitions = new ArrayList<>(topicIdPartitions);
+        // We want the elements from the end of the list to move left by the distance provided i.e. if the original list is [1,2,3],
+        // and we want to rotate it by 1, we want the output as [2,3,1] and not [3,1,2]. Hence, we need negation of distance here.
```
```suggestion
        // Elements from the list should move left by the distance provided i.e. if the original list is [1,2,3],
        // and rotation is by 1, then output should be [2,3,1] and not [3,1,2]. Hence, negate the distance here.
```

apoorvmittal10 commented on server/src/test/java/org/apache/kafka/server/share/fetch/PartitionRotateStrategyTest.java:
```diff
@@ -35,64 +36,64 @@ public class PartitionRotateStrategyTest {
 
     /**
-     * Create an ordered map of TopicIdPartition to partition max bytes.
+     * Create an ordered set of topic partitions.
```
```suggestion
     * Create a list of topic partitions.
```

apoorvmittal10 commented on server/src/test/java/org/apache/kafka/server/share/fetch/PartitionRotateStrategyTest.java:
```diff
@@ -35,64 +36,64 @@ public class PartitionRotateStrategyTest {
+     * Create an ordered set of topic partitions.
      * @param size The number of topic-partitions to create.
-     * @return The ordered map of TopicIdPartition to partition max bytes.
+     * @return The ordered set of topic partitions.
```
```suggestion
     * @return The list of topic partitions.
```

apoorvmittal10 reviewed (commented):
LGTM, some minor comments and a question.

adixitconfluent committed:
Addressed Apoorv's review comments round 3

apoorvmittal10 reviewed (approved):
Thanks for the changes, LGTM.
